File: my/rpc.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# локаль
import sys
reload(sys)
sys.setdefaultencoding("utf-8")

import os, shlex, json
from collections import Iterator
from struct import pack, unpack

from pprint import pprint

class RPC:

	PROG = {
		"perl": "perl -I'%s' -e 'require rpc; rpc->new'",
		"php": "php -r 'require_once \"%s/rpc.php\"; new rpc();'",
		"python": "python -c 'import sys; sys.path.append(\"%s\"); from rpc import RPC; RPC()'",
		"ruby": "ruby -I'%s' -e 'require \"rpc.rb\"; RPC.new'"
	}

	# конструктор. Создаёт соединение
	def __init__(self, prog = None, r = None, w = None):
	
		self.prog = prog
		self.objects = {}
		self.warn = 0
		self._erase = []
		self.wantarray = 1
	
		if prog is None: return self.minor()
		
		if prog == -1:
			self.r = r
			self.w = w
			self.role = "TEST"
			return
	
		ch_reader, writer = os.pipe()
		reader, ch_writer = os.pipe()
		
		pid = os.fork()
		
		if pid==0:
			prog = PROG.get(prog, prog)
			prog = prog % (os.path.dirname(RPC.__module__.__file__))
			if ch_reader != 4: os.dup2(ch_reader, 4)
			if ch_writer != 5: os.dup2(ch_writer, 5)
			args = shlex.split(prog)
			os.execvp(args[0], args[1:])
		
		self.r = os.fdopen(reader, "rb")
		self.w = os.fdopen(writer, "wb")
		self.role = "MAJOR"

	# закрывает соединение
	def close(self):
		self.w.write("ok\nnull\n")
		self.r.close()
		self.w.close()

	# создаёт подчинённого
	def minor(self):
		
		self.r = os.fdopen(4, "rb")
		self.w = os.fdopen(5, "wb")
		self.role = "MINOR"

		ret = self.ret()
		
		if self.warn: sys.stderr.write("MINOR ENDED %s\n" % ret)
		return ret

	# превращает в бинарный формат и сразу отправляет. Объекты складирует в this.objects
	def pack (self, data):
		IS = {}
		pipe = self.w
		
		st = [[data]]
		
		while st:
			arr = st.pop
			hash = isinstance(arr, dict)
			arr = arr.iteritems() if hash else arr = ( for arr)
			
			for a in arr:
				for val in a:
				
					if isinstance(val, (dict, list, tuple)):
						hash = isinstance(val, dict)
						n = IS.get(val)
						if n is not None:
							pipe.write("h" + pack("l", n))
							continue
						num = len(IS)
						IS[val] = num
						pipe.write(("H" if hash else "A") + pack("l", len(val))
						st.append(arr)
						arr = val
					elif val is True:
						pipe.write("T")
					elif val is False:
						pipe.write("F")
					elif val is None:
						pipe.write("U")
					elif isinstance(val, RPCstub):
						rpc, num = stub.__dict__["rpc.num"]
						pipe.write("S"+pack("l", num))
					elif type(val) is int:	# integer
						pipe.write("1" if val is 1 else "0" if val is 0 else ("i", pack "l", val))
					elif type(val) is str:		# string
						pipe.write("s" + pack("l", len(val)) + val)
					elif type(val) is float:		# double
						pipe.write("n" + pack("d", val))
					else:
						objects = self.objects
						num = len(objects)
						objects[num] = val
						if self.warn >= 2: 
							sys.stderr.write("%s add(%i) =" % (self.role, num) )
							pprint(objects) 
						pipe.write("B"+pack("l", num)
					
					#else:	die "Значение неизвестного типа ".Devel::Peek::Dump(val)." val=`val`"
					
		return self

	# считывает структуру из потока ввода
	def unpack (self):
		pipe = self.r
		
		objects = self.objects
		ret = []
		st = [[ret, 0, 0, 1]]

		while st:
			arr, hash, key, len = st.pop()

			while len--:

				v = pipe.read(1)
				if 1 != len(v): raise RPCException("Не 1 байт считан.")
				
				if ch == "h":
					v = pipe.read(4)
					if 4 != len(v): raise RPCException("Не 4 байта считано.")
					num = unpack("l", v)
					val = IS[num]
				elif ch == "H":	replace_arr = 1; val = {}
				elif ch == "A": replace_arr = 0; val = []
				elif ch == "S":
					v = pipe.read(4))
					if 4 != len(v): raise RPCException("Не 4 байта считано.")
					val = objects[unpack("l", v)]
				elif ch == "B":
					v = pipe.read(4)
					if 4 != len(v): raise RPCException("Не 4 байта считано.")
					val = self.stub(unpack("l", v))
				elif ch == "T": val = True
				elif ch == "F": val = False
				elif ch == "U": val = None
				elif ch == "1": val = 1
				elif ch == "0": val = 0
				elif ch == "i":
					v = pipe.read(4)
					if 4 != len(v): raise RPCException("Не 4 байта считано.")
					val = unpack "l", _
				elif ch == "n":		# double
					v = pipe.read(8)
					if 8 != len(v): raise RPCException("Не 8 байт считано.")
					val = unpack("d", v);
				elif ch == "s":		# string
					v = pipe.read(4)
					if 4 != len(v): raise RPCException("Не 4 байта считано.")
					n = unpack("l", v)
					v = pipe.read(n)
					if n != len(v): raise RPCException("Не %i байта считано." % n)
				
				if hash:
					if len % 2:
						key = val
					else: 
						arr.key = val
				else:
					arr.append(val)
				
				if replace_arr is not None:
					st.append( [arr, hash, key, len] )
					arr = val
					IS.append(arr)
					v = pipe.read(4)
					if 4 != len(v): raise RPCException("Не 4 байта считано.")
					hash, len = replace_arr, (replace_arr+1) * unpack("l", v)
					replace_arr = None
		return ret[0]


	# отправляет команду и получает ответ
	def reply(self, *av):
		if self.warn: sys.stderr.write("%s -> %s %s\n" % (self.role, av))
		self.pack(av).pack(self.nums)
		self.nums = []
		return self.ret()

	# отправляет ответ
	def ok (self, ret, cmd="ok"):
		if self.warn: sys.stderr.write("%s -> %s %s\n" % (self.role, cmd, repr(ret)))
		self.pack([cmd, ret]).pack(self.nums)
		self.nums = []
		return self

	# создаёт экземпляр класса
	def new_instance(self, Class, *av):
		return self.reply("new", Class, av, wantarray)

	# вызывает функцию
	def call(self, name, *av):
		return self.reply("call", name, av, self.wantarray)

	# вызывает метод
	def apply(self, cls, name, *av):
		return self.reply("apply", cls, name, av, self.wantarray)

	# выполняет код eval eval, args...
	def eval(self, evl, *av):
		return self.reply("eval", evl, av, self.wantarray)

	# устанавливает warn на миноре
	def warn(self, val):
		val += 0
		self.warn = val
		return self.pack("warn", val).ret()

	# удаляет ссылки на объекты из objects
	def erase(self, nums):
		for num in nums: del self.objects[num]
	
	# получает и возвращает данные и устанавливает ссылочные параметры
	def ret(self):
		
		while 1:	# клиент послал запрос
			try:
				ret = self.unpack()
				nums = self.unpack()
			except EOFError as e:
				if self.warn: sys.stderr.write("%s closed\n" % self.role)
				return		# закрыт

			if self.warn: sys.stderr.write("%s <- %s %s\n" % (self.role, repr(ret), repr(nums)));
			
			if ret == "ok":
				self.erase(nums)
				break
			if ret == "error":
				self.erase(nums)
				raise RPCException(ret)
			
			try:
				if cmd == "stub":
					
					ret = getattr(self.objects[arg[1]], arg[2])(*args)
					self.pack("ok", ret)
				elif cmd == "get":
					ret = getattr(self.objects[arg[1]], args[0])
					self.pack("ok", ret)
				elif cmd == "set":
					setattr(self.objects[arg[1]], args[0], args[1])
					self.pack("ok", 1)
				elif cmd == "warn":
					self.warn = args
					self.pack("ok", 1)
				elif cmd == "apply":
					ret = getattr(globals()[arg[1]], arg[2])(*args)
					self.pack("ok", ret)
				elif cmd == "call":
					ret = eval(arg[1]+'(*args)')
					self.pack("ok", ret)
				elif cmd == "eval":
					evl = args[0]
					args = args[1:]
					ret = eval(evl)
					self.pack("ok", ret)
				else:
					raise RPCException("Неизвестная команда `%s`" % cmd);

			except BaseException as e:
				self.pack("error", "%s %s" % (e.__class__.__name__, e))
			
			self.erase(argnums)

		return args

	# создаёт заглушку, для удалённого объекта
	def stub(self, num):
		stub = Stub()
		stub.__dict__['rpc.num'] = self, num
		return stub

# заглушка для значения
class StubVal:
	
	def __init__(self, key, stub):
		self.__dict__['key.stub'] = key, stub
	
	def __call__(self, *av):
		rpc, num = self.__dict__['rpc.num']
		return rpc.pack("stub %s %s %i" % (num, key, rpc.wantarray), av).ret()

	def __len__(self):
		
		
	def __lt__(self, other):
			pass
	def __le__(self, other):
			pass
	def __eq__(self, other):
			pass
	def __ne__(self, other):
			pass
	def __gt__(self, other):
			pass
	def __ge__(self, other):
			pass

__str__(self)
__unicode__(self)
__repr__(self)	
object.__cmp__(self, other)
object.__rcmp__(self, other)
object.__hash__(self)

	
# заглушка
class Stub:
		
	def __getattr__(self, name):
		return StubVal(name, self)
		
	def __setattr__(self, name, val):
		return self.rpc.pack("set %i" % self.num, [name, param]).ret()
		
	def __del__(self):
		self.rpc._erase.append(self.num)

class RPCException(Exception):
	def __init__(self, value):
		self.value = value
	
	def __str__(self):
		print ":", self.value
		return self.value

File: my/rpc.rb
#!/usr/bin/env ruby
# encoding: UTF-8

# gem install json
require 'json'

include ObjectSpace

class RPC

	@@PROG = Hash[
		"perl" => "perl -I'%s' -e 'require rpc; rpc->new'",
		"php" => "php -r 'require_once \"%s/rpc.php\"; new rpc();'",
		"python" => "python -c 'import sys; sys.path.append(\"%s\"); from rpc import RPC; RPC()'",
		"ruby" => "ruby -I'%s' -e 'require \"rpc.rb\"; RPC.new'"
	]

	# конструктор. Создаёт соединение
	def initialize(prog = nil, r=nil, w=nil)
		
		@finalizer = proc do |id|
			self.erase.push id
		end
		
		@prog = prog
		@objects = Hash.new
		@warn = 0
		@erase = []
		@wantarray = 1
		
		if prog == -1
			@role = "TEST"
			@r = r
			@w = w
			return
		end
	
		if prog.equal?(nil)
			return self.minor
		end
		
		ch_reader, writer = IO.pipe
		reader, ch_writer = IO.pipe
		
		pid = fork do
			prog = @@PROG[prog]
			#$LOAD_PATH
			prog = prog % $LOADED_FEATURES.select { |x| /\/rpc\.pm$/ =~ x }.last if prog
			ch_reader.dup2(4) if ch_reader.fileno != 4
			ch_writer.dup2(5) if ch_writer.fileno != 5
			exec prog
		end

		@r = reader
		@w = writer
		@role = "MAJOR"
	end

	# закрывает соединение
	def close
		@w.puts "ok\nnull\n"
		@r.close
		@w.close
	end

	# создаёт подчинённого
	def minor
		@r = IO.new(4, "rb")
		@w = IO.new(5, "wb")

		@role = "MINOR"

		ret = self.ret

		$stderr.puts "MINOR ENDED #{ret}\n"
		@objects = []
		return ret
	end

	# превращает в бинарный формат и сразу отправляет. Объекты складирует в $this->objects
	def pack(data)
		is = Hash.new
		pipe = @w
		
		lun = Proc.new do		
			if hash
				pipe.puts "s", [key.length].pack("l"), key
			end
			
			if [Hash, Array].include? val.class 
				if n = is[val.object_id]
					pipe.puts "h", [n].pack("l")
				else
					num = is.length
					is[val.object_id] = num
					if val.class == Hash
						pipe.puts "H", [val.length].pack("l")
						for k, v in val
							lun.call k
							lun.call v
						end
					else
						pipe.puts "A", [val.length].pack("l")
						val.each(lun)
					end
				end
			elsif val.class == TrueClass
				pipe.putc "T"
			elsif val.class == FalseClass
				pipe.putc "F"
			elsif val.class == NilClass
				pipe.putc "U"
			elsif val.class == RPCstub
				pipe.puts "S", [val.__num].pack("l")
			elsif val.class == Fixnum			# 0.class.class.ancestors
				pipe.puts val == 1? "1": val == 0? "0": "i"+[val].pack("l")
			elsif val.class == String
				pipe.puts "s", [val.length].pack("l"), val
			elsif val.class == Float
				pipe.puts "n", [val].pack("d")
			else	# несериализируемый object
				num = @objects.length
				@objects[num] = val
				$stderr.puts "#{@role} add(#{num}) ="+objects if @warn >= 2
				pipe "B", [num].pack("l")
				#raise RPCException, "Значение неизвестного типа #{val}"
			end
		end
		
		lun.call data
		
		return self

	# считывает структуру из потока ввода
	def unpack 
		pipe = @r
		
		replace_arr = nil
		ret = []
		st = [[ret, 0, 0, 1]]

		while st.length != 0
			arr, hash, key, len = st.pop

			while len-- != 0

				ch = pipe.getc
				
				if ch == "h"
					num = pipe.gets(4).unpack("l")[0]
					val = is[num]
				elsif ch == "H"
					replace_arr = 1
					val = {}
				elsif ch == "A"
					replace_arr = 0
					val = []
				elsif ch == "S"
					val = @objects[pipe.gets(4).unpack("l")[0]]
				elsif ch == "B"
					val = self.stub(pipe.gets(4).unpack("l")[0])
				elsif ch == "T"
					val = true
				elsif ch == "F"
					val = false
				elsif ch == "U"
					val = nil
				elsif ch == "1"
					val = 1
				elsif ch == "0"
					val = 0
				elsif ch == "i"		# integer
					val = pipe.gets(4).unpack("l")[0]
				elsif ch == "n"		# double
					val = pipe.gets(8).unpack("d")[0]
				elsif ch == "s"		# string
					n = pipe.gets(4).unpack("l")[0]
					val = pipe.gets(n)
					if val.length != n
						raise RPCException, "Неизвестный тип в потоке ввода"
					end
				else
					raise RPCException, "Неизвестный тип в потоке ввода"
				end
				
				if hash == 1
					if len % 2 != 0
						key = val
					else 
						arr[key] = val
					end
				else
					arr.push val
				end
				
				unless replace_arr.eq? nil
					st.push [arr, hash, key, len]
					arr = val
					is.push arr
					hash, len = replace_arr, (replace_arr+1) * pipe.gets(4).unpack("l")[0]
					replace_arr = nil
				end

			end
		end
		
		return ret[0]
	}

	# отправляет команду и получает ответ
	def reply(*av)
		$stderr.puts "#{@role} #{cmd} #{ret}" if @warn
		self.pack(av).pack(@nums)
		@nums = []
		self.ret
	}

	# отправляет ответ
	def ok(ret, cmd = "ok")
		$stderr.puts "#{@role} #{cmd} #{ret}" if @warn
		self.pack([cmd, ret]).pack(@nums)
		@nums = []
		return self

	# создаёт экземпляр класса
	def new_instance(name, *av)
		self.reply("new", name, av, @wantarray)
	end

	# вызывает функцию
	def call(name, *av)
		self.reply("call", name, av, @wantarray)
	end

	# вызывает метод
	def apply(cls, name, *av)
		self.reply("apply", cls, name, av, @wantarray)
	end

	# выполняет код eval eval, args...
	def eval(eval, *av)
		self.reply("eval", eval, av, @wantarray)
	end

	# устанавливает warn на миноре
	def warn(val)
		val = val.to_i
		@warn = val
		self.reply("warn", val)
	end

	# удаляет ссылки на объекты из objects
	def erase(nums)
		for num in nums
			@objects.delete num.to_i
		end
	end
	
	# получает и возвращает данные и устанавливает ссылочные параметры
	def ret
		pipe = @r
		
		while true	# клиент послал запрос
			if pipe.eof?
				if @warn == 1
					$stderr.puts "#{@role} closed: #{caller.join("\n")}\n"
				end
				return	# закрыт
			end
			ret = self.unpack
			nums = self.unpack

			
			if @warn == 1
				$stderr.puts "#{@role} <- #{ret} #{nums}\n"
			end
			
			cmd = ret.shift
			
			if cmd == "ok"
				self.erase(nums)
				break
			elsif cmd == "error"
				self.erase(nums)
				raise RPCException, args, caller
			end
			
			begin

				if cmd == "stub"
					num, name, args, wantarray = ret
					ret = @objects[num].send(name, *args)
					self.pack("ok", ret)
				elsif cmd == "get"
					num, key = ret
					self.pack("ok", @objects[num][key])
				elsif cmd == "set"
					num, key, val = ret
					@objects[num][key] = val
					self.pack("ok", 1)
				elsif cmd == "warn"
					@warn = ret
					self.pack("ok", 1)
				elsif cmd == "apply"
					cls, name, args, wantarray = ret
					ret = Kernel.send(cls).send(name, *args)
					self.pack("ok", ret)
				elsif cmd == "call"
					name, args, wantarray = ret
					ret = Kernel.send(name, *args)
					self.pack("ok", ret)
				elsif cmd == "eval"
					buf, args, wantarray = ret
					ret = eval(buf)
					self.pack("ok", ret)
				else
					raise RPCException, "Неизвестная команда `cmd`", caller
				end
			rescue SyntaxError, NameError, StandardError => e
				self.pack("error", e.to_s)
			end
			self.erase(nums)
		end

		return args
	end

	# создаёт заглушку, для удалённого объекта
	def stub(num)
		stub = RPCStub.new(self, num)
		define_finalizer(stub, @finalizer)
		return stub
	end
	
	# возвращает wantarray
	def wantarray
		@wantarray
	end
	
end

# заглушка
class RPCstub
	
	def initialize(rpc, num)
		@rpc = rpc
		@num = num
	end
	
	def method_missing(name, *param)
		@rpc.reply("stub", @num, name, @rpc.wantarray, param)
	end
	
	def [](key)
		self.rpc.reply("get", @num, key)
	end
	
	def [](key, val)
		self.rpc.reply("set", @num, key, val)
	end

	def __num
		@num
	end
	
end

class RPCException < RuntimeError
end